gcc/rust/checks/errors/borrowck/rust-bir.h

```diff
@@ -0,0 +1,200 @@
+// Copyright (C) 2020-2023 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#ifndef RUST_BIR_BASE_H
+#define RUST_BIR_BASE_H
+
+#include "rust-bir-place.h"
+#include "rust-bir-visitor.h"
+
+namespace Rust {
+
+namespace BIR {
+
+struct BasicBlock;
+class Node;
+class AbstractExpr;
+
+/**
+ * Top-level entity of the Borrow-checker IR (BIR).
+ * It represents a single function (method, closure, etc.), which is also the
+ * basic unit of Polonius borrow-checking.
+ */
+struct Function
+{
+  PlaceDB place_db;
+  std::vector<PlaceId> arguments; // Only used for dump.
+  std::vector<BasicBlock> basic_blocks;
+};
+
+/** Single "instruction" of BIR. */
+class Node
+{
+public:
+  enum Kind
+  {
+    ASSIGNMENT,	  // <place> = <expr>
+    SWITCH,	  // switch <place>
+    RETURN,	  // return
+    GOTO,	  // goto
+    STORAGE_DEAD, // StorageDead(<place>)
+    STORAGE_LIVE, // StorageLive(<place>)
+  };
+
+private:
+  Kind kind;
+  // ASSIGNMENT: lhs
+  // SWITCH: switch_val
+  // StorageDead/StorageLive: place
+  // otherwise: <unused>
+  PlaceId place;
+  // ASSIGNMENT: rhs
+  // otherwise: <unused>
+  std::unique_ptr<AbstractExpr> expr;
+
+public:
+  Node (PlaceId lhs, AbstractExpr *rhs)
+    : kind (Kind::ASSIGNMENT), place (lhs), expr (rhs)
+  {}
+
+  explicit Node (Kind kind, PlaceId place = INVALID_PLACE,
+		 AbstractExpr *expr = nullptr)
+    : kind (kind), place (place), expr (expr)
+  {}
+
+public:
+  [[nodiscard]] Kind get_kind () const { return kind; }
+  [[nodiscard]] PlaceId get_place () const { return place; }
+  [[nodiscard]] AbstractExpr &get_expr () const { return *expr; }
```

Comment:
we need to handle `[[nodiscard]]` nicely in our CI and in the GCC build system for this to work properly, otherwise the compiler will complain that it is a C++17 extension (despite having the proper behavior in C++11 mode). @tschwinge @dkm do you know if this is used in other frontends/how to deal with it?

Reply:
I take that citation back, that requirement is C++17.

Reply:
The standard way would be to define a version/feature-dependent macro in `gcc/system.h`.  For example, like the one that already exists in `libstdc++-v3/include/bits/c++config`:

```c++
// Macro to warn about unused results.
#if __cplusplus >= 201703L
# define _GLIBCXX_NODISCARD [[__nodiscard__]]
#else
# define _GLIBCXX_NODISCARD
#endif
```

However, can't you just use the existing `gcc/system.h:WARN_UNUSED_RESULT` (see <https://gcc.gnu.org/onlinedocs/gcc-4.8.5/gcc/Function-Attributes.html#index-g_t_0040code_007bwarn_005funused_005fresult_007d-attribute-2785>):

```c++
#if GCC_VERSION >= 3004
#define WARN_UNUSED_RESULT __attribute__ ((__warn_unused_result__))
#else
#define WARN_UNUSED_RESULT
#endif
```

Or does that have different semantics than `nodiscard`?

Reply:
Replaced for WARN_UNUSED_RESULT.